src/imap/mod.rs

```diff
@@ -1245,19 +1297,78 @@ fn parse_message_id(message_id: &[u8]) -> crate::error::Result<String> {
     bail!("could not parse message_id: {}", value);
 }
 
-fn prefetch_get_message_id(prefetch_msg: &Fetch) -> Result<String> {
-    if prefetch_msg.envelope().is_none() {
-        return Err(Error::Other(
-            "prefectch: message has no envelope".to_string(),
-        ));
+fn get_fetch_headers(prefetch_msg: &Fetch) -> Result<Vec<mailparse::MailHeader>> {
+    let header_bytes = match prefetch_msg.header() {
+        Some(header_bytes) => header_bytes,
+        None => return Ok(Vec::new()),
+    };
+    let (headers, _) = mailparse::parse_headers(header_bytes)?;
+    Ok(headers)
+}
+
+fn prefetch_get_message_id(headers: &[mailparse::MailHeader]) -> Result<String> {
+    if let Some(message_id) = headers.get_first_value(&HeaderDef::MessageId.get_headername())? {
+        Ok(parse_message_id(&message_id)?)
+    } else {
+        Err(Error::Other("prefetch: No message ID found".to_string()))
     }
+}
+
+/// Checks if fetch result contains a header
+fn prefetch_has_header(headers: &[mailparse::MailHeader], headerdef: HeaderDef) -> Result<bool> {
+    Ok(headers
+        .get_first_value(&headerdef.get_headername())?
+        .is_some())
+}
 
-    let message_id = prefetch_msg.envelope().unwrap().message_id;
-    if message_id.is_none() {
-        return Err(Error::Other("prefetch: No message ID found".to_string()));
+fn prefetch_is_reply_to_chat_message(
+    context: &Context,
+    headers: &[mailparse::MailHeader],
+) -> Result<bool> {
+    if let Some(value) = headers.get_first_value(&HeaderDef::InReplyTo.get_headername())? {
+        if is_msgrmsg_rfc724_mid_in_list(context, &value) {
+            return Ok(true);
+        }
     }
 
-    parse_message_id(&message_id.unwrap()).map_err(Into::into)
+    if let Some(value) = headers.get_first_value(&HeaderDef::References.get_headername())? {
+        if is_msgrmsg_rfc724_mid_in_list(context, &value) {
+            return Ok(true);
+        }
+    }
+
+    Ok(false)
+}
+
+fn prefetch_should_download(
+    context: &Context,
+    headers: &[mailparse::MailHeader],
+    show_emails: ShowEmails,
+) -> Result<bool> {
+    let is_chat_message = prefetch_has_header(&headers, HeaderDef::ChatVersion)?;
+    let is_autocrypt_setup_message =
+        prefetch_has_header(&headers, HeaderDef::AutocryptSetupMessage)?;
+
+    // Autocrypt Setup Message should be shown even if it is from non-chat client.
```

Comment:
this comment needs to move up above the let.

Reply:
Fixed